fileshare/event_manager.go

```diff
@@ -66,17 +70,60 @@ func NewEventManager(
 	filesystem Filesystem,
 	defaultDownloadDir string,
 ) *EventManager {
-	return &EventManager{
+	em := &EventManager{
 		isProd:                isProd,
 		liveTransfers:         map[string]*LiveTransfer{},
 		transferSubscriptions: map[string]chan TransferProgressInfo{},
 		meshClient:            meshClient,
 		osInfo:                osInfo,
 		filesystem:            filesystem,
 		defaultDownloadDir:    defaultDownloadDir,
+		syncEvents:            make(chan []Event),
+		syncDoneCh:            make(chan struct{}),
+		asyncEvents:           make(chan []Event, 2048),
+	}
+	go em.process()
+
+	return em
+}
+
+func (em *EventManager) process() {
+	fn := func(ev []Event) {
+		em.mutex.Lock()
+		for _, e := range ev {
+			em.handleEvent(e)
+		}
+		em.mutex.Unlock()
+	}
+
+	for {
+		select {
+		case e, ok := <-em.asyncEvents:
+			if !ok {
+				log.Println(internal.WarningPrefix + " asyncEvents channel closed")
+				return
+			}
+			fn(e)
+		case e, ok := <-em.syncEvents:
+			if !ok {
+				log.Println(internal.WarningPrefix + " syncEvents channel closed")
+				return
+			}
+			fn(e)
+			em.syncDoneCh <- struct{}{}
+		}
 	}
 }
 
+func (em *EventManager) AsyncEvent(event ...Event) {
+	em.asyncEvents <- event
+}
+
+func (em *EventManager) SyncEvent(event ...Event) {
```

Comment:
So this is used only in tests?

Reply:
Basically yes, but I don't want to put it to test-only file, because it can be used in production code as well.

Reply:
I think I would go with YAGNI rule here :smile:, because we don't have any signals that synchronous processing will be needed. IMHO it's better to not try to foresee the future, it works in very rare cases

What do you think?

Reply:
@devzbysiu but then what do you do with the need for having synchronous processing in tests? Maybe I can just make the `SyncEvent()` function private?